osu.Game/Overlays/FirstRunSetup/ScreenImportFromStable.cs

```diff
@@ -0,0 +1,268 @@
+
+                countUpdateCancellation?.Cancel();
+                countUpdateCancellation = new CancellationTokenSource();
+
+                legacyImportManager.GetImportCount(StableContent, countUpdateCancellation.Token).ContinueWith(task => Schedule(() =>
+                {
+                    if (task.IsCanceled)
+                        return;
+
+                    int count = task.GetResultSafely();
+
+                    LabelText = LocalisableString.Interpolate($"{title} ({FirstRunOverlayImportFromStableScreenStrings.Items(count)})");
+                }));
+            }
+        }
+
+        internal class StableLocatorLabelledTextBox : LabelledTextBoxWithPopover, ICanAcceptFiles
+        {
+            [Resolved]
+            private LegacyImportManager legacyImportManager { get; set; } = null!;
+
+            // TODO: test
+            public IEnumerable<string> HandledExtensions { get; } = new[] { string.Empty };
+
+            private readonly Bindable<DirectoryInfo> currentDirectory = new Bindable<DirectoryInfo>();
+
+            [Resolved]
+            private OsuGameBase game { get; set; } = null!;
+
+            protected override void LoadComplete()
+            {
+                base.LoadComplete();
+
+                game.RegisterImportHandler(this);
+
+                currentDirectory.BindValueChanged(onDirectorySelected);
+
+                string? fullPath = legacyImportManager.GetCurrentStableStorage()?.GetFullPath(string.Empty);
+                if (fullPath != null)
+                    currentDirectory.Value = new DirectoryInfo(fullPath);
+            }
+
+            private void onDirectorySelected(ValueChangedEvent<DirectoryInfo> directory)
+            {
+                if (directory.NewValue == null)
+                {
+                    Current.Value = string.Empty;
+                    return;
+                }
+
+                // DirectorySelectors can trigger a noop value changed, but `DirectoryInfo` equality doesn't catch this.
+                if (directory.OldValue?.FullName == directory.NewValue.FullName)
+                    return;
+
+                if (directory.NewValue?.GetFiles(@"osu!.*.cfg").Any() ?? false)
+                {
+                    this.HidePopover();
+
+                    string path = directory.NewValue.FullName;
+
+                    legacyImportManager.UpdateStorage(path);
+                    Current.Value = path;
+                }
+            }
+
+            Task ICanAcceptFiles.Import(params string[] paths)
+            {
+                Schedule(() => currentDirectory.Value = new DirectoryInfo(paths.First()));
+                return Task.CompletedTask;
+            }
+
+            Task ICanAcceptFiles.Import(params ImportTask[] tasks) => throw new NotImplementedException();
+
+            protected override void Dispose(bool isDisposing)
+            {
+                base.Dispose(isDisposing);
+                game.UnregisterImportHandler(this);
```

Comment:
A touch unsafe if disposal ever gets called before injection of `game`. Normally would have been nullchecked but that's a bit awkward here...

Reply:
I don't think this can happen, as we don't finalise drawables and if this was added to a hierarchy it would be non-null, right?